Feat: PP_Queue_v1 & PP_Queue_ManualExecution

vendrell46 opened the pull request:
## What has been done?
- Setup of PP_Queue, PP_Queue_ManualExecution, interfaces
- Manage queue based on the client
- Add functions to
  - add payment orders to the queue
  - execute the queue
  - cancel individual payment orders in the queue, removing the element form the queue
- add functions to validate an a payment order for the PP_Queue
- Extend PP_Queue to add a manual execution function

## Relevant Links
- [Gherkin Specification](https://hackmd.io/M2BhprupTyeZ1u4l4HeSgw?view#Payment-Processor)

vendrell46 committed:
Contract structure defined

vendrell46 committed:
removed IFundingManager extension

vendrell46 committed:
IFundingManager_v1 needed

vendrell46 committed:
extended IRedeemingBondingCurveBase_v1 as well

vendrell46 committed:
ExternalPriceSetter, blacklist, Oracle and interfaces added

vendrell46 committed:
formatting files and adapting to the code of conduct

vendrell46 committed:
ManualExternalPriceSetter_v1 modified as LM

vendrell46 committed:
added fees checks, paymentOrder to the process and defined role for blacklist

vendrell46 committed:
managing roles, renaming, and removing unneeded things

vendrell46 committed:
whitelist logic removed, blacklist fixes

vendrell46 committed:
missing logic implemented

vendrell46 committed:
blacklist fixes and custom error added

vendrell46 committed:
switched from role to mapping for blacklist manager

vendrell46 committed:
moved blacklist files to external, modified sell, and more context added

vendrell46 committed:
PAYMENT_QUEUE_ROLE added, _sellOrder overriden

scab24 committed:
emit_timestamp(IssuancePriceSet,RedemptionPriceSet),NatSpec

Zitzak committed:
fix: remove collateral transfer & collateral check

- The function `handleCollateralTokensAfterSell` was still expecting to
transfer collateral, which would revert in the current implementation.
- The _sellOrder` function had a conditional check which would let
the contract revert given there will not be any collateral in the
contract

Zitzak committed:
style: run make fmt for formating

vendrell46 committed:
small fixes and documentation added

vendrell46 committed:
docs added for issuance and redemption setters

vendrell46 committed:
Rename ERC20Issuance_blacklist_v1.sol to ERC20Issuance_Blacklist_v1.sol

vendrell46 committed:
Rename IERC20Issuance_blacklist_v1.sol to IERC20Issuance_Blacklist_v1.sol

vendrell46 committed:
fixed executeRedemptionQueue, docs and renaming

vendrell46 committed:
formatting natspec

Zitzak committed:
feat: add treasury collateral withdraw

Zitzak committed:
Fix fmt

Zitzak committed:
fix: typo

Zitzak committed:
fix: fmt

Zitzak committed:
Merge pull request #2 from ZealynxSecurity/add_collateral_transfer

Feat: Add collateral transfer to treasury

Zitzak committed:
fix: remove oracle from init

Zitzak committed:
Merge pull request #3 from ZealynxSecurity/removeOracleAddress

fix: remove oracle from init

vendrell46 committed:
fixed natspec

vendrell46 committed:
modifier moved

Zitzak committed:
feat: update openRedemptionAmount

Zitzak committed:
Merge pull request #4 from ZealynxSecurity/update_open_redemption_amount

FHieser committed:
Fix: Remove redeemed amount from balance check

* Remove unnecessary safety check

* Abstract projectFeeCollected

* Create tests

* Fmt

* Update src/modules/fundingManager/bondingCurve/interfaces/IRedeemingBondingCurveBase_v1.sol

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>

* Add check to _handleCollateralTokensAfterSell

* Revert "Add check to _handleCollateralTokensAfterSell"

This reverts commit 32f9bb353784c05975e43c69d3006de8c79e8b58.

* Readd the check but only for projectCollateralFeeCollected

* Update IFM_BC_Bancor_Redeeming_VirtualSupply_v1.sol

* fmt

* Apply comment suggestions from code review

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Rename Error

* Apply suggestions from code review

* Fix: Resolve version mismatches

* Chore: Resolve formatting issue

---------

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>
Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

FHieser committed:
Fix: Remove redeemed amount from balance check

* Remove unnecessary safety check

* Abstract projectFeeCollected

* Create tests

* Fmt

* Update src/modules/fundingManager/bondingCurve/interfaces/IRedeemingBondingCurveBase_v1.sol

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>

* Add check to _handleCollateralTokensAfterSell

* Revert "Add check to _handleCollateralTokensAfterSell"

This reverts commit 32f9bb353784c05975e43c69d3006de8c79e8b58.

* Readd the check but only for projectCollateralFeeCollected

* Update IFM_BC_Bancor_Redeeming_VirtualSupply_v1.sol

* fmt

* Apply comment suggestions from code review

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Rename Error

* Apply suggestions from code review

* Fix: Resolve version mismatches

* Chore: Resolve formatting issue

---------

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>
Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
applied suggestions

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
formatting

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
formatting

vendrell46 committed:
Merge branch 'dev-1' of https://github.com/ZealynxSecurity/InverterContracts into dev-1

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
formatting

scab24 committed:
Modifier buyingIsEnabled and sellingIsEnabled added

vendrell46 committed:
formatting, updating code from parent and adding interface check

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

Zitzak committed:
initial setup

vendrell46 committed:
formatting

vendrell46 committed:
formatting

vendrell46 committed:
moved constants

vendrell46 committed:
changed notice to dev

vendrell46 committed:
adding named returned variable

vendrell46 committed:
modifier renamed and buyingIsEnabled added to buy

vendrell46 committed:
return named variables to interface

vendrell46 committed:
formatting

Zitzak committed:
fix: revert during deployment

Zitzak committed:
Merge pull request #6 from ZealynxSecurity/fix_setting_error

fix: revert during deployment

Zitzak committed:
fix issue with authorization

Zitzak committed:
Merge pull request #7 from ZealynxSecurity/add_internal_setter_function

fix issue with authorization

Zitzak committed:
fix: interface check

Zitzak committed:
Merge pull request #8 from ZealynxSecurity/fix_interface_check

fix: interface check

vendrell46 committed:
interface natspec same as contract

vendrell46 committed:
line too long fixed

vendrell46 committed:
removed dot from title

vendrell46 committed:
_ensureTokenBalance overriden

Zitzak committed:
remove function

Zitzak committed:
Merge branch 'dev-1' into remove_function

Zitzak committed:
fix fmt

Zitzak committed:
Merge pull request #9 from ZealynxSecurity/remove_function

remove function from FM

vendrell46 committed:
work started with PP_Queue_v1

Zitzak committed:
Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

Zitzak committed:
fix: Move LM_ManualPriceSetter into LM folder

Zitzak committed:
Merge pull request #10 from ZealynxSecurity/move_lm_to_right_folder

fix: Move LM_ManualPriceSetter into LM folder

vendrell46 committed:
Feat: FM_PC_ExternalPriceRedeeming, ManualExternalPriceSetter and ERC20Issuance_Blacklist

* Contract structure defined

* removed IFundingManager extension

* IFundingManager_v1 needed

* extended IRedeemingBondingCurveBase_v1 as well

* ExternalPriceSetter, blacklist, Oracle and interfaces added

* formatting files and adapting to the code of conduct

* ManualExternalPriceSetter_v1 modified as LM

* added fees checks, paymentOrder to the process and defined role for blacklist

* managing roles, renaming, and removing unneeded things

* whitelist logic removed, blacklist fixes

* missing logic implemented

* blacklist fixes and custom error added

* switched from role to mapping for blacklist manager

* moved blacklist files to external, modified sell, and more context added

* PAYMENT_QUEUE_ROLE added, _sellOrder overriden

* emit_timestamp(IssuancePriceSet,RedemptionPriceSet),NatSpec

* fix: remove collateral transfer & collateral check

- The function `handleCollateralTokensAfterSell` was still expecting to
transfer collateral, which would revert in the current implementation.
- The _sellOrder` function had a conditional check which would let
the contract revert given there will not be any collateral in the
contract

* style: run make fmt for formating

* small fixes and documentation added

* docs added for issuance and redemption setters

* Rename ERC20Issuance_blacklist_v1.sol to ERC20Issuance_Blacklist_v1.sol

* Rename IERC20Issuance_blacklist_v1.sol to IERC20Issuance_Blacklist_v1.sol

* fixed executeRedemptionQueue, docs and renaming

* formatting natspec

* feat: add treasury collateral withdraw

* Fix fmt

* fix: typo

* fix: fmt

* fix: remove oracle from init

* fixed natspec

* modifier moved

* feat: update openRedemptionAmount

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* applied suggestions

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* formatting

* Update src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* formatting

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* formatting

* Modifier buyingIsEnabled and sellingIsEnabled added

* formatting, updating code from parent and adding interface check

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* formatting

* formatting

* moved constants

* changed notice to dev

* adding named returned variable

* modifier renamed and buyingIsEnabled added to buy

* return named variables to interface

* formatting

* fix: revert during deployment

* fix issue with authorization

* fix: interface check

* interface natspec same as contract

* line too long fixed

* removed dot from title

* _ensureTokenBalance overriden

* remove function

* fix fmt

* Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* fix: Move LM_ManualPriceSetter into LM folder

---------

Co-authored-by: scab24 <git.seco@protonmail.com>
Co-authored-by: Zitzak <marvin88gr@gmail.com>
Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>
Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>
Co-authored-by: Sergio <94926493+scab24@users.noreply.github.com>

vendrell46 committed:
Merge remote-tracking branch 'upstream/feature/PP_Queue_v1' into pp_queue_initialSetup

vendrell46 committed:
PP_Queue_v1 implemented

Zitzak committed:
fix: decimal conversion error

* fix: decimal conversion error

* fix: add 10^tokenDecimal

Zitzak committed:
fix: decimal conversion error

* fix: decimal conversion error

* fix: add 10^tokenDecimal

vendrell46 committed:
store unclaimable amounts after transfer failure

vendrell46 committed:
changed to queue per client system

Zitzak committed:
Feat: Add deduction of openRedemptionAmount

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

Zitzak committed:
Feat: Add deduction of openRedemptionAmount

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

scab24 committed:
FM_PC_ExternalPrice_Redeeming_v1_Test

vendrell46 committed:
removing unnecessary things and adding quick fixes

vendrell46 committed:
fixing compilation error

vendrell46 committed:
Update src/modules/paymentProcessor/PP_Queue_ManualExecution_v1.sol

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>

vendrell46 committed:
Update src/modules/paymentProcessor/PP_Queue_ManualExecution_v1.sol

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>

vendrell46 committed:
vesting constants removed

vendrell46 committed:
reused _validPaymentOrder

vendrell46 committed:
formatting and reverting call to _validPaymentOrder

Zitzak committed:
Merge pull request #5 from ZealynxSecurity/pp_queue_initialSetup

PP Queue inital setup

Zitzak commented:
@marvinkruse this was one open comment in the other PR, which I'm linking here
https://github.com/ZealynxSecurity/InverterContracts/pull/5#discussion_r1895526781

scab24 committed:
setup

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+            order.order.amount
+        );
+
+        // Remove from queue if present
```
```suggestion
        // Remove from queue if present.
```

vendrell46 replied:
done

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        }
+
+        uint processedCount;
+        while (firstId != LinkedIdList._SENTINEL && _processNextOrder(client_))
```
Is there no way to call `_processNextOrder` without going through this `while` loop (i.e. a from-to-approach)? I am just thinking of the case of someone spamming the queue with 1wei withdraw requests only to have this always run out of gas? Or am I missing something?

vendrell46 replied:
I see the risk that you're mentioning. Do you have any suggestion how to implement this differently that it's also optimal and not too complex?

I'm not sure now what could be the workaround for this, @marvinkruse 

cc. @Zitzak

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+
+        queueId_ = _getPaymentQueueId(order_.flags, order_.data);
+
+        // Create new order
```
```suggestion
        // Create new order.
```

vendrell46 replied:
done

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+            client: client_
+        });
+
+        // Add to linked list
```
```suggestion
        // Add to linked list.
```

vendrell46 replied:
done

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        _queue[client_].removeId(prevId, orderId_);
+    }
+
+    /// @dev    Validates token balance and allowance for a payment.
```
Should be a `notice` tag.

vendrell46 replied:
done

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+    }
+
+    /// @notice	Used to claim the unclaimable amount of a particular
+    /// paymentReceiver for a given payment client.
```
Indentation went wrong here.

vendrell46 replied:
done

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+    /// @param	client Address of the payment client.
+    /// @param	token Address of the payment token.
+    /// @param	paymentReceiver Address of the paymentReceiver for which
+    /// the unclaimable amount will be claimed.
```
Indentation went wrong here.

vendrell46 replied:
done

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        emit TokensReleased(paymentReceiver, address(token), amount);
+    }
+
+    /// @dev    Validates if a queue ID is valid.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        return queueId_ > 0 && queueId_ <= _nextOrderId;
+    }
+
+    /// @dev    Gets payment queue ID from flags and data.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+    {
+        // Check if orderID flag is set (bit 0)
+        bool hasOrderId = uint(flags_) & (1 << FLAG_ORDER_ID) != 0;
+        if (hasOrderId) {
```
We only care about the happy-path here, does it not matter if the order id was not transmitted?

vendrell46 replied:
covered other cases

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        }
+    }
+
+    /// @dev    Validate uint total amount input.
```
Should be a `notice` tag. Also "uint total amount" sounds a bit weird.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        );
+    }
+
+    /// @dev    Validates a payment token.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        }
+    }
+
+    /// @dev    Validates a payment order.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        view
+        returns (bool valid_)
+    {
+        // Validate recipient
```
Missing dots at the end here in this func.

vendrell46 replied:
added

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        }
+
+        // Validate amount
+        if (!_validTotalAmount(order_.order.amount)) {
```
Already validated 10 lines prior.

vendrell46 replied:
removed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        return true;
+    }
+
+    /// @dev    Validates a state transition.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        return true;
+    }
+
+    /// @dev    Updates the state of a payment order.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        emit PaymentOrderStateChanged(orderId_, state_, order.client);
+    }
+
+    /// @dev    Validates flags and corresponding data array.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        }
+    }
+
+    /// @dev    Internal function to check whether the client is valid.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        }
+    }
+
+    /// @dev    Sets the unclaimable amount for a specific payment.
```
Should be a `notice` tag.

vendrell46 replied:
changed

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        _updateOrderState(orderId_, RedemptionState.CANCELLED);
+
+        // Add amount to unclaimable
+        _setUnclaimableAmount(
```
This is overriding the currently stored amount for that user (for that token + client). Should this be added or is it intentional to fully override it, even in case the user has multiple orders?

vendrell46 replied:
hey good catch! You're right. I've modified it to add on top of the existing amount and renamed function to `_addToUnclaimableAmount`

marvinkruse commented on src/modules/paymentProcessor/interfaces/IPP_Queue_ManualExecution_v1.sol:
```diff
@@ -0,0 +1,44 @@
+    // Functions
+
+    /// @notice  Executes all pending payment orders in the queue.
+    /// @dev     This function processes all pending orders in the queue.
```
This dev tag adds zero information and is unncessary.
```suggestion
```

marvinkruse commented on src/modules/paymentProcessor/interfaces/IPP_Queue_v1.sol:
```diff
@@ -0,0 +1,233 @@
+// Internal
+import {IPaymentProcessor_v1} from "@pp/IPaymentProcessor_v1.sol";
+import {IERC20PaymentClientBase_v1} from
+    "../../logicModule/interfaces/IERC20PaymentClientBase_v1.sol";
```
Should use `@lm` I think

vendrell46 replied:
you're right

marvinkruse commented on src/modules/paymentProcessor/PP_Queue_ManualExecution_v1.sol:
```diff
@@ -0,0 +1,101 @@
+import {IOrchestrator_v1} from
+    "src/orchestrator/interfaces/IOrchestrator_v1.sol";
+import {IPaymentProcessor_v1} from
+    "src/modules/paymentProcessor/IPaymentProcessor_v1.sol";
```
Should use `@pp`

vendrell46 replied:
changed

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+    // Constants
+
+    /// @dev    Role for queue operations.
+    bytes32 public constant QUEUE_OPERATOR_ROLE = "QUEUE_OPERATOR";
```
I've added the `getQueueOperatorRole` function

Zitzak replied:
Where did this role go? 🥲 @vendrell46

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+    }
+
+    /// @inheritdoc IPaymentProcessor_v1
+    function unclaimable(address client, address token, address paymentReceiver)
```
added

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+    }
+
+    /// @inheritdoc IPaymentProcessor_v1
+    function claimPreviouslyUnclaimable(
```
added

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+    // Constants
+
+    /// @dev    Role for queue operations.
+    bytes32 public constant QUEUE_OPERATOR_ROLE = "QUEUE_OPERATOR";
```
I've added the getQueueOperatorRole function

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+    }
+
+    /// @inheritdoc IPaymentProcessor_v1
+    function unclaimable(address client, address token, address paymentReceiver)
```
added

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+            _unclaimableAmountsForRecipient[client][token][paymentReceiver];
+    }
+
+    /// @inheritdoc IPaymentProcessor_v1
```
yes, added

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+        _unclaimableAmountsForRecipient[client_][token_][receiver_] = amount_;
+    }
+
+    /// @dev    Checks if an order exists.
```
changed

vendrell46 commented on src/modules/paymentProcessor/interfaces/IPP_Queue_v1.sol:
```diff
@@ -0,0 +1,233 @@
+    /// @param  client_ Address of the client whose queue failed.
+    error Module__PP_Queue_QueueOperationFailed(address client_);
+
+    /// @notice	Caller not authorized for operation.
```
changed

vendrell46 committed:
renaming and formatting

vendrell46 committed:
Update src/modules/paymentProcessor/PP_Queue_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/paymentProcessor/PP_Queue_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/paymentProcessor/PP_Queue_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Update src/modules/paymentProcessor/interfaces/IPP_Queue_ManualExecution_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+            => mapping(
+                address token => mapping(address receiver => uint amount)
+            )
+    ) private _unclaimableAmountsForRecipient;
```
modified

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+
+    /// @notice Tracks all payments that could not be made to the
+    ///         paymentReceiver.
+    /// @dev    client => token => receiver => unclaimable amount.
```
removed

vendrell46 committed:
Update src/modules/paymentProcessor/PP_Queue_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,740 @@
+            revert Module__PP_Queue_InvalidState();
+        }
+
+        // Update order state and emit event
```
done

vendrell46 commented on src/modules/paymentProcessor/interfaces/IPP_Queue_v1.sol:
```diff
@@ -0,0 +1,233 @@
+    // -------------------------------------------------------------------------
+    // Structs
+
+    /// @notice	Queued payment order information.
```
added them

vendrell46 committed:
formatting and dots

vendrell46 committed:
underscores

vendrell46 committed:
handling unhappy paths to get queue id

vendrell46 committed:
make fmt

scab24 committed:
new_test_PP_Queue

scab24 committed:
new_test_PP_Queue

vendrell46 committed:
fixing nextOrderId and so on

scab24 committed:
Update PP_Queue_v1.sol

scab24 committed:
Update IPP_Queue_v1.sol

scab24 committed:
new_test

scab24 committed:
new_tests

vendrell46 committed:
new tests added

vendrell46 committed:
more tests for exposed functions added

vendrell46 committed:
audit fixes

scab24 committed:
Tests_PP_Queue

scab24 committed:
Fix_Audit + Tests

Zitzak committed:
refactor: remove duplicate ERC20Mock, extend existing with decimals

Zitzak committed:
refactor: initial cleanup LM_ManualExternalPriceSetter_v1

- removed unused imports
- moved files to right directory
- renamed files
- cleaned up setup by removing unnecessary initialization, instead
reuse existing setup from ModuleTest
- format code (make fmt)

Zitzak committed:
refactor: initial setup ERC20Issuance_Blacklist_v1
- Moved file
- Removed unnecessary imports
- Use only exposed mock for testing

Zitzak committed:
refactor: FM_PC_ExternalPrice_Redeeming_v1_Test

- Remove imports
- Refactor most of the setup
- Add exposed mock
- Removed OraclePriceMock and InvalidMock
- Added simpler oracle mockf
- fmt

Zitzak committed:
refactor: remove import and fmt of LM_ManualExternalPriceSetter test

leeftk committed:
refactor:changed file structure

leeftk committed:
fix:fee amount bug

leeftk committed:
refactor:add internal func for redeem and issue

Zitzak committed:
Merge pull request #12 from ZealynxSecurity/updateTesting

Refactor: Initial refactor of testing

Zitzak committed:
fix: compile error

scab24 committed:
Fix=> testExternalBuy_revertsGivenNonWhitelistedUser, testExternalBuy_succeedsGivenMaxAmount and comment testExternalQueue_succeedsGivenSingleRedemptionOrder

scab24 committed:
Add format (Inverter Standard) PP_Queue, make pre-commit

leeftk committed:
fix:projectfeeAmount variable

leeftk committed:
audit fix:remove normalize prices

leeftk committed:
fix:failing test for issuance decimals

leeftk committed:
more audit fixes

0xNuggan committed:
address _setFlags() comment

Zitzak committed:
refactor(ERC20Issuance_Blacklist): Update event, internal function

What has been done:
- Add msg.sender to events
- Move logic of modifier to internal function (Inverter Standard)

Zitzak committed:
refactor(ERC20Issuance_Blacklist_v1_Test):

- Cleaned setup
- Moved imports and functions to correct place (Inverter Standard)
- Added extensive gherkin
- Add some missing tests

vendrell46 committed:
Merge feature/FM_PC_Oracle_Redeeming_v1 with audit fixes

Resolved conflicts while preserving audit fixes:
- Removed _nextOrderId and related functions
- Simplified price handling in ManualExternalPriceSetter
- Updated event definitions
- Fixed function visibility and modifiers

Zitzak committed:
Merge branch 'dev-1' into refactor_testing_USP_FM

Zitzak committed:
feat(FM_PC_ExternalPrice_Redeeming): add new paymennt order struct

Zitzak committed:
Merge remote-tracking branch 'upstream/feature/FM_PC_Oracle_Redeeming_v1' into ZealynxSecurity/Feature/FM_PC_ExternalPrice

leeftk committed:
fix:add missing events to interface

leeftk committed:
fix:inherit doc issue, code compiling, tests pass

leeftk committed:
fix:remove unused exposed contracts

Zitzak committed:
Merge branch 'ZealynxSecurity/Feature/FM_PC_ExternalPrice' into refactor_testing_USP_FM

Zitzak committed:
Merge pull request #14 from ZealynxSecurity/refactor_testing_USP_FM

Refactor testing usp fm

leeftk committed:
refactor: refactor set issuance and redemption price

leeftk committed:
refacotr: use internal functions and remove redundant function

leeftk committed:
add:deductFromRedemption function

leeftk committed:
fix:add msg.sender to price setter,move events to ioracle

Zitzak committed:
docs(IOraclePrice_v1): add missing param natspec

Zitzak committed:
feat: Add roles for the role admin

vendrell46 committed:
event with contract address and add amountPaid added

vendrell46 committed:
Merge branch 'feature/PP_Queue_v1' into feat/PP_Queue

vendrell46 committed:
emit event that updates state with msgSender

vendrell46 committed:
merging events and fixing compilation error

scab24 committed:
Add PP_Queue.sol and IPP_Queue.sol

scab24 committed:
Add FundingManagerPaymentProcessorE2E, add configuration in E2EModuleRegistry, tests(2)

scab24 committed:
improve setup, add _setupOrchestratorFundingManagerPaymentProcessor

vendrell46 committed:
format - 1st round

vendrell46 committed:
compilation errors fixed

vendrell46 committed:
fixed 2 tests out of 16

vendrell46 committed:
formatting - 2nd stage

leeftk committed:
format: update imports and rearrange public functions

leeftk committed:
format:change content structure

scab24 committed:
fix decimals in _redeemTokensFormulaWrapper

Zitzak committed:
client order ID fix

scab24 committed:
deposit funds, Execute Queue

scab24 committed:
Fix tests, delete console.log, Deposit funds in preparation for executing Queue, Execute Queue

Zitzak committed:
cherry pick from other branch

Zitzak committed:
Merge branch 'test-pp-queue_v1' into feat/PP_Queue

scab24 committed:
fix (;)

Zitzak commented on src/modules/paymentProcessor/PP_Queue_v1.sol:
```diff
@@ -0,0 +1,774 @@
+    /// @notice Validates a payment order.
+    /// @param  order_ The order to validate.
+    /// @return valid_ True if the order is valid.
+    function _validPaymentOrder(QueuedOrder memory order_)
```
The function called within the internal function are also not utilized, like the `_validateFlagsAndData()`

vendrell46 committed:
Merge branch 'Zealynx-e2e_tests_FM+PP' into feat/PP_Queue

vendrell46 committed:
bringing back queue verification and compilation error fixes

vendrell46 committed:
fixed: testClaimPreviouslyUnclaimable_GivenMultipleAmounts

vendrell46 committed:
fixed testCancelPaymentOrder_RevertGivenZeroId:

- _orderExists would return true as it was even if the order didn't exist (which had orderId = 0)

- Now we verify:
   1. The order belongs to the specified client (order.client_ == address(client_))
   2. The order has been initialized (order.timestamp_ != 0)

vendrell46 committed:
fixed `testGetQueueHead_RevertGivenUninitializedQueue` by:

- Checking if the queue is initialized by looking at the sentinel position in the queue's list mapping
- If the sentinel position is 0, it means the queue is uninitialized, so we revert with our own error
- Only if the queue is initialized, we proceed to get the head of the queue

Zitzak committed:
refactor(PP_Queue):

What has been done:
- Add Queue operator, queue operator admin role and getters
- refactor internal _validPaymentOrder() and call the function
  in different funcitons to replace individual validation statements
- Remove timestamp from event emit
- Update natspec across different functions
- Refactor _validateFlagsAndData() to return boolean instead of revert
- Remove unused imports

Zitzak committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/PP_Queue' into feat/PP_Queue

Zitzak committed:
refactor(PP_Queue): Fix open questions around todos

What has been done:
- Remove duplicate check for 0 order ID
- Change revert in _validPaymentToken() to return boolean instead

Zitzak committed:
format

Zitzak committed:
test: fix (some) PP_queue tests

Zitzak committed:
merge up to date FM and adjecent contracts into PP_Queue branch

Zitzak committed:
WIP

scab24 committed:
FIX test_PP_Queue

scab24 committed:
test(e2e): fix setup and role assignments

- Fix main configuration and role assignments in E2E tests
- Reorganize preconditions for better test structure
- Fix and enable 8 test cases in QueueBaseFundingManagerAndPaymentProcessorE2E
- Improve role assignment logic with correct module addresses
- Clean up test initialization and setup process
- Ensure proper role hierarchy in _setadminroles function
- Format code

Zitzak committed:
remove canceled order state

Zitzak committed:
refactor(PP_Queue_v1): add features

What has been done?
- Add addresses, getters and setters for cancelled and failed order
  treasuries
- fix _orders mapping to work with multiple clients
- Add function to claim unclaimableAmount from blacklisted address
  to the right treasury
- Refactor cancelPaymentOrderThroughQueueId
- Refactor _executePaymentTransfer such that transfers are handled
  seperately, as transfers are needed across the module

Zitzak committed:
merge FM

Zitzak committed:
test: update tests

Zitzak commented:
This branch got merged into the ZealynxSecurity/Feature/FM_PC_ExternalPrice feature branch, for which a PR is open [here](https://github.com/InverterNetwork/contracts/pull/712)
